File: src-tauri/src/commands/query.rs
//! Ejecución de SQL escrito por el usuario.
//!
//! Cada pestaña del editor abre su propia conexión y la conserva mientras siga abierta. Los
//! resultados no se devuelven de una: viajan por un canal a medida que cada sentencia termina, con
//! el mismo mecanismo que usa el mantenimiento. Un script de diez sentencias tiene que mostrar la
//! primera cuando termina la primera, no cuando terminan las diez.

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Instant;

use pgforge_core::error::ErrorPayload;
use pgforge_core::sql::{
    self, completion, explain, history::Entry as HistoryEntry, history::Source as HistorySource,
    ColumnType, ExplainOptions, Limits, NewEntry, NewQuery, Outcome, Plan, QuerySession,
    SavedQuery, SchemaSnapshot, TxStatus,
};
use pgforge_core::{Error, ProfileId, Result};
use serde::Serialize;
use tauri::ipc::Channel;
use tauri::State;
use tokio::sync::Mutex;

use crate::state::{AppState, QueryEntry};

#[derive(Clone, Serialize)]
#[serde(
    tag = "type",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum QueryEvent {
    /// Empezó una sentencia del script.
    #[serde(rename_all = "camelCase")]
    Started {
        index: usize,
        total: usize,
        line: usize,
    },
    /// La sentencia devolvió filas. Va en `Box` porque es mucho más grande que el resto y no tiene
    /// sentido que cada evento pague ese tamaño.
    #[serde(rename_all = "camelCase")]
    Finished {
        index: usize,
        outcome: Box<Outcome>,
    },
    Notice {
        severity: String,
        message: String,
    },
    /// Falló una sentencia; las que quedaban no se ejecutan.
    #[serde(rename_all = "camelCase")]
    Failed {
        index: usize,
        error: ErrorPayload,
        /// Dónde empieza la sentencia dentro del script, en caracteres. Sumado a la posición que
        /// trae el error, ubica el carácter exacto en el editor.
        offset: usize,
    },
    /// Cómo quedó la transacción de la pestaña. Se emite al terminar el script, haya fallado o no:
    /// una sentencia rota dentro de una transacción la deja abortada, y eso hay que mostrarlo.
    #[serde(rename_all = "camelCase")]
    Transaction {
        status: TxStatus,
    },
    /// Terminó el script entero.
    #[serde(rename_all = "camelCase")]
    Completed {
        seconds: f64,
        /// Cuántas sentencias llegaron a ejecutarse.
        executed: usize,
    },
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryTab {
    pub tab_id: String,
    /// La base que quedó abierta, que puede ser la del perfil si no se pidió otra.
    pub database: String,
    /// Valor heredado del perfil; la pestaña lo puede cambiar con `query_autocommit`.
    pub autocommit: bool,
    pub tx_status: TxStatus,
}

/// Abre la conexión de una pestaña de consulta.
#[tauri::command]
pub async fn query_open(
    state: State<'_, AppState>,
    id: ProfileId,
    database: Option<String>,
) -> Result<QueryTab> {
    let handle = state.manager.require(id).await?;
    let database = database.unwrap_or_else(|| handle.default_database().to_owned());

    let mut session = QuerySession::open(&handle, &database).await?;
    let cancel = session.cancel_token();
    let notices = session.take_notices();
    let sink = Arc::new(Mutex::new(None));
    let tab_id = uuid::Uuid::new_v4().to_string();

    // Los NOTICE llegan por la conexión en cualquier momento, no como respuesta a una consulta:
    // una tarea de por vida los reparte a la ejecución que esté en curso, si es que hay alguna.
    if let Some(mut notices) = notices {
        let sink: Arc<Mutex<Option<Channel<QueryEvent>>>> = sink.clone();
        tokio::spawn(async move {
            while let Some(notice) = notices.recv().await {
                if let Some(channel) = sink.lock().await.as_ref() {
                    let _ = channel.send(QueryEvent::Notice {
                        severity: notice.severity,
                        message: notice.message,
                    });
                }
            }
        });
    }

    let autocommit = handle.profile.autocommit;

    state.queries.lock().await.insert(
        tab_id.clone(),
        QueryEntry {
            profile: id,
            database: database.clone(),
            session: Arc::new(Mutex::new(session)),
            cancel,
            notices: sink,
            autocommit: Arc::new(AtomicBool::new(autocommit)),
        },
    );

    Ok(QueryTab {
        tab_id,
        database,
        autocommit,
        // Una sesión recién abierta no tiene nada empezado; no hace falta preguntárselo al servidor.
        tx_status: TxStatus::Idle,
    })
}

/// Cierra la pestaña y suelta su conexión.
#[tauri::command]
pub async fn query_close(state: State<'_, AppState>, tab_id: String) -> Result<()> {
    state.queries.lock().await.remove(&tab_id);
    Ok(())
}

/// Ejecuta un script sentencia por sentencia.
///
/// Los errores viajan por el canal y no como falla de la llamada: la interfaz necesita saber
/// *cuál* de las sentencias falló y dónde, y eso no entra en un `Err` genérico.
#[tauri::command]
pub async fn query_run(
    state: State<'_, AppState>,
    tab_id: String,
    sql: String,
    limits: Option<Limits>,
    channel: Channel<QueryEvent>,
) -> Result<()> {
    let (profile, database, session, sink, autocommit) = {
        let tabs = state.queries.lock().await;
        let entry = require(&tabs, &tab_id)?;
        (
            entry.profile,
            entry.database.clone(),
            entry.session.clone(),
            entry.notices.clone(),
            entry.autocommit.load(Ordering::Relaxed),
        )
    };

    let statements = sql::split(&sql);
    if statements.is_empty() {
        return Err(Error::Config(
            "no hay ninguna sentencia para ejecutar".to_owned(),
        ));
    }

    let limits = limits.unwrap_or_default();
    let started = Instant::now();
    let started_at = epoch_seconds();
    let mut executed = 0;
    let mut rows_returned = None;
    let mut failure = None;

    // La sesión queda tomada durante todo el script: es una sola conexión, y dejar que dos
    // ejecuciones se intercalen mezclaría las transacciones de la pestaña.
    let session = session.lock().await;
    *sink.lock().await = Some(channel.clone());

    // Con autocommit apagado, la transacción la abre pgforge antes de la primera sentencia: el
    // servidor no tiene modo autocommit que apagar. Si ya había una abierta, esto no hace nada.
    if !autocommit {
        if let Err(error) = session.begin_if_needed().await {
            *sink.lock().await = None;
            return Err(error);
        }
    }

    for (index, statement) in statements.iter().enumerate() {
        let _ = channel.send(QueryEvent::Started {
            index,
            total: statements.len(),
            line: statement.line,
        });

        match session.run(&statement.text, limits).await {
            Ok(outcome) => {
                executed += 1;
                if let Outcome::Rows { row_count, .. } = &outcome {
                    rows_returned = Some(*row_count as i64);
                }
                let _ = channel.send(QueryEvent::Finished {
                    index,
                    outcome: Box::new(outcome),
                });
            }
            Err(error) => {
                let _ = channel.send(QueryEvent::Failed {
                    index,
                    error: ErrorPayload::from(&error),
                    offset: statement.offset,
                });
                failure = Some(error.to_string());
                break;
            }
        }
    }

    // Cómo quedó la transacción se pregunta acá y no se deduce del SQL: el usuario pudo escribir su
    // propio `BEGIN`, un `COMMIT` o dejarla abortada, y la interfaz tiene que mostrar lo que hay.
    // Que falle la sonda no invalida lo que ya se ejecutó, así que solo se omite el evento.
    if let Ok(status) = session.tx_status().await {
        let _ = channel.send(QueryEvent::Transaction { status });
    }

    *sink.lock().await = None;
    let seconds = started.elapsed().as_secs_f64();

    // Se registra lo que el usuario mandó a ejecutar, no cada sentencia por separado: es lo que va
    // a querer recuperar del historial, y es lo que reconoce cuando lo ve.
    let recorded = state.history.lock().await.record(&NewEntry {
        source: HistorySource::Editor,
        profile_id: profile.to_string(),
        database,
        sql,
        started_at,
        seconds,
        row_count: rows_returned,
        error: failure,
    });
    // Que falle el historial no puede tirar abajo una consulta que ya se ejecutó: se avisa junto a
    // los demás mensajes y la ejecución se da por buena igual.
    if let Err(error) = recorded {
        let _ = channel.send(QueryEvent::Notice {
            severity: "WARNING".to_owned(),
            message: error.to_string(),
        });
    }

    let _ = channel.send(QueryEvent::Completed { seconds, executed });

    Ok(())
}

/// Segundos desde el epoch. Un reloj corrido hacia atrás daría un instante negativo; se recorta a
/// cero antes que romper el registro por algo que no importa.
fn epoch_seconds() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|elapsed| elapsed.as_secs() as i64)
        .unwrap_or_default()
}

/// Últimas consultas ejecutadas, de un servidor o de todos.
#[tauri::command]
pub async fn history_recent(
    state: State<'_, AppState>,
    id: Option<ProfileId>,
    limit: Option<i64>,
) -> Result<Vec<HistoryEntry>> {
    let profile = id.map(|id| id.to_string());
    state
        .history
        .lock()
        .await
        .recent(profile.as_deref(), limit.unwrap_or(100))
}

#[tauri::command]
pub async fn history_search(
    state: State<'_, AppState>,
    text: String,
    limit: Option<i64>,
) -> Result<Vec<HistoryEntry>> {
    state
        .history
        .lock()
        .await
        .search(&text, limit.unwrap_or(100))
}

#[tauri::command]
pub async fn history_clear(state: State<'_, AppState>) -> Result<()> {
    state.history.lock().await.clear()
}

/// Las consultas guardadas, por nombre.
#[tauri::command]
pub async fn saved_list(state: State<'_, AppState>) -> Result<Vec<SavedQuery>> {
    state.saved.lock().await.list()
}

/// Guarda una consulta nueva o reescribe una existente. El reloj lo pone acá quien tiene el sistema
/// operativo, no el núcleo, igual que en el historial.
#[tauri::command]
pub async fn saved_save(state: State<'_, AppState>, query: NewQuery) -> Result<SavedQuery> {
    state.saved.lock().await.save(&query, epoch_seconds())
}

#[tauri::command]
pub async fn saved_delete(state: State<'_, AppState>, saved_id: i64) -> Result<bool> {
    state.saved.lock().await.delete(saved_id)
}

/// Pide al servidor que aborte lo que la pestaña esté ejecutando.
#[tauri::command]
pub async fn query_cancel(state: State<'_, AppState>, tab_id: String) -> Result<()> {
    let (profile, cancel) = {
        let tabs = state.queries.lock().await;
        let entry = require(&tabs, &tab_id)?;
        (entry.profile, entry.cancel.clone())
    };

    // Se le pide al servidor, no se aborta la tarea local: matarla del lado de la aplicación
    // dejaría al servidor ejecutando la consulta igual, sin nadie esperando el resultado.
    let handle = state.manager.require(profile).await?;
    handle.cancel(&cancel).await
}

/// Confirma la transacción de la pestaña y devuelve cómo quedó.
#[tauri::command]
pub async fn query_commit(state: State<'_, AppState>, tab_id: String) -> Result<TxStatus> {
    let session = session_of(&state, &tab_id).await?;
    let session = session.lock().await;

    // El estado se consulta aunque el `COMMIT` falle —una restricción diferida puede voltearlo—,
    // porque la interfaz tiene que mostrar la transacción real, no la que se esperaba.
    let applied = session.commit().await;
    let status = session.tx_status().await;
    applied?;
    status
}

/// Revierte la transacción de la pestaña, esté sana o abortada.
#[tauri::command]
pub async fn query_rollback(state: State<'_, AppState>, tab_id: String) -> Result<TxStatus> {
    let session = session_of(&state, &tab_id).await?;
    let session = session.lock().await;

    let applied = session.rollback().await;
    let status = session.tx_status().await;
    applied?;
    status
}

/// Los tipos de las columnas que devolvería la sentencia, sin ejecutarla.
///
/// Se pide aparte porque cuesta preparar la sentencia en el servidor: la interfaz lo hace solo
/// cuando el usuario pide ver los tipos.
#[tauri::command]
pub async fn query_column_types(
    state: State<'_, AppState>,
    tab_id: String,
    sql: String,
) -> Result<Vec<ColumnType>> {
    let session = session_of(&state, &tab_id).await?;
    let session = session.lock().await;
    session.column_types(&sql).await
}

/// Estado de la transacción de la pestaña, para volver a sincronizar la interfaz.
#[tauri::command]
pub async fn query_tx_status(state: State<'_, AppState>, tab_id: String) -> Result<TxStatus> {
    let session = session_of(&state, &tab_id).await?;
    let session = session.lock().await;
    session.tx_status().await
}

/// Enciende o apaga el autocommit de la pestaña.
///
/// Encenderlo **no confirma nada**: solo deja de anteponer el `BEGIN`. Si quedaba una transacción
/// abierta se devuelve tal cual está, para que la interfaz pida confirmarla o revertirla — commitear
/// en silencio por tocar un interruptor es exactamente lo que estos botones existen para evitar.
#[tauri::command]
pub async fn query_autocommit(
    state: State<'_, AppState>,
    tab_id: String,
    enabled: bool,
) -> Result<TxStatus> {
    let (session, autocommit) = {
        let tabs = state.queries.lock().await;
        let entry = require(&tabs, &tab_id)?;
        (entry.session.clone(), entry.autocommit.clone())
    };

    autocommit.store(enabled, Ordering::Relaxed);
    let session = session.lock().await;
    session.tx_status().await
}

/// Sesión de una pestaña, para los comandos que solo necesitan eso.
async fn session_of(state: &State<'_, AppState>, tab_id: &str) -> Result<Arc<Mutex<QuerySession>>> {
    let tabs = state.queries.lock().await;
    Ok(require(&tabs, tab_id)?.session.clone())
}

/// Plan de ejecución de una sentencia.
#[tauri::command]
pub async fn query_explain(
    state: State<'_, AppState>,
    tab_id: String,
    sql: String,
    options: Option<ExplainOptions>,
) -> Result<Plan> {
    let session = {
        let tabs = state.queries.lock().await;
        require(&tabs, &tab_id)?.session.clone()
    };

    let options = options.unwrap_or_default();
    let session = session.lock().await;

    // Del script se explica una sola sentencia: un plan por sentencia no es un plan.
    let statements = sql::split(&sql);
    let statement = statements
        .first()
        .ok_or_else(|| Error::Config("no hay ninguna sentencia para explicar".to_owned()))?;

    explain::explain(&session, &statement.text, options).await
}

/// Nombres del esquema para el autocompletado. La interfaz lo pide una vez por pestaña.
#[tauri::command]
pub async fn schema_snapshot(
    state: State<'_, AppState>,
    id: ProfileId,
    database: Option<String>,
) -> Result<SchemaSnapshot> {
    let handle = state.manager.require(id).await?;
    let database = database.unwrap_or_else(|| handle.default_database().to_owned());

    completion::snapshot(&handle, &database).await
}

/// La sentencia sobre la que está el cursor, para poder ejecutar solo esa.
///
/// Lo resuelve el núcleo y no la interfaz: reimplementar en TypeScript las reglas de dónde termina
/// una sentencia daría dos partidores que se van a desincronizar el día que aparezca un caso raro.
#[tauri::command]
pub fn statement_at_cursor(sql: String, cursor: usize) -> Option<sql::Statement> {
    sql::at_cursor(&sql, cursor)
}

/// Guarda el texto de una pestaña de consulta en la ruta elegida.
///
/// Misma excepción anotada que `erd_export_svg`: lo que se escribe lo tiene la interfaz y no el
/// servidor, así que no hay lógica que llevar al núcleo, y estas dos líneas salen más baratas que
/// sumar el complemento de archivos con su ámbito de permisos.
#[tauri::command]
pub fn sql_write_file(path: String, sql: String) -> Result<()> {
    std::fs::write(path, sql)?;
    Ok(())
}

/// Lee un archivo de SQL para abrirlo en una pestaña. Misma excepción que el de arriba.
///
/// El texto se lee con pérdida (`from_utf8_lossy`) en vez de fallar: un script viejo guardado en
/// Latin-1 tiene que poder abrirse y arreglarse, y negarse a mostrarlo por un acento roto no ayuda
/// a nadie.
#[tauri::command]
pub fn sql_read_file(path: String) -> Result<String> {
    let bytes = std::fs::read(path)?;
    Ok(String::from_utf8_lossy(&bytes).into_owned())
}

/// Advertencia previa a un `EXPLAIN ANALYZE`, que ejecuta la sentencia de verdad.
///
/// Es una consulta al núcleo y no una regla escrita en la interfaz para que la CLI avise lo mismo.
#[tauri::command]
pub fn explain_warning(sql: String, options: Option<ExplainOptions>) -> Option<String> {
    explain::warning(&sql, options.unwrap_or_default()).map(str::to_owned)
}

fn require<'a>(
    tabs: &'a std::collections::HashMap<String, QueryEntry>,
    tab_id: &str,
) -> Result<&'a QueryEntry> {
    tabs.get(tab_id)
        .ok_or_else(|| Error::Config("la pestaña de consulta ya no está abierta".to_owned()))
}

File: ui/src/lib/ipc/query.ts
import { Channel } from "@tauri-apps/api/core";

import { invoke, type CoreError } from "./core";

// ---------------------------------------------------------------------------
// Consultas
// ---------------------------------------------------------------------------

export interface QueryTab {
  tabId: string;
  database: string;
  autocommit: boolean;
  txStatus: TxStatus;
}

/** Estado de la transacción de una pestaña. `failed` es «abierta pero abortada». */
export type TxStatus = "idle" | "active" | "failed";

export interface QueryLimits {
  maxRows: number;
}

/** Resultado de una sentencia. Refleja `pgforge_core::sql::Outcome`. */
export type Outcome =
  | {
      kind: "rows";
      columns: string[];
      /** `null` es un NULL de la base, distinto de la cadena vacía. */
      rows: (string | null)[][];
      rowCount: number;
      truncated: boolean;
      seconds: number;
    }
  | { kind: "command"; tag: string; affected: number; seconds: number };

export type QueryEvent =
  | { type: "started"; index: number; total: number; line: number; offset: number }
  | { type: "finished"; index: number; outcome: Outcome }
  | { type: "notice"; severity: string; message: string }
  | { type: "failed"; index: number; error: CoreError; offset: number }
  | { type: "transaction"; status: TxStatus }
  | { type: "completed"; seconds: number; executed: number };

export interface ExplainOptions {
  analyze: boolean;
  buffers: boolean;
  verbose: boolean;
}

export interface PlanNode {
  nodeType: string;
  relation: string | null;
  /** El esquema de la relación; llega porque el plan se pide con VERBOSE. */
  schema: string | null;
  index: string | null;
  condition: string | null;
  /** El `Filter` suelto: en un `Index Scan`, lo que el índice no resolvió. */
  filter: string | null;
  startupCost: number;
  totalCost: number;
  planRows: number;
  actualRows: number | null;
  loops: number | null;
  totalMs: number | null;
  /** Tiempo del nodo sin el de sus hijos: el que señala al culpable. */
  selfMs: number | null;
  rowsRemoved: number | null;
  misestimated: boolean;
  sharedHitBlocks: number | null;
  sharedReadBlocks: number | null;
  sortMethod: string | null;
  sortSpaceKb: number | null;
  /** El orden no entró en `work_mem` y terminó escribiéndose en disco. */
  sortOnDisk: boolean;
  children: PlanNode[];
}

export type AdviceKind = "missingIndex" | "indexFilter" | "staleStats" | "workMem";

/** Lo que conviene mirar del plan. Nunca se aplica solo: la sentencia se copia (ver `sql::advice`). */
/** Sobre qué tabla y por qué columnas se propone el índice, ya desarmado para el diálogo. */
export interface IndexTarget {
  schema: string;
  table: string;
  columns: string[];
}

export interface Advice {
  kind: AdviceKind;
  severity: "warn" | "info";
  title: string;
  detail: string;
  sql: string | null;
  index: IndexTarget | null;
}

export interface Plan {
  root: PlanNode;
  planningMs: number | null;
  executionMs: number | null;
  analyzed: boolean;
  advice: Advice[];
  /** La respuesta del servidor tal cual, para pegarla en un visor de planes. */
  json: string;
}

export const queryOpen = (id: string, database?: string) =>
  invoke<QueryTab>("query_open", { id, database: database ?? null });

export const queryClose = (tabId: string) => invoke<void>("query_close", { tabId });

export const queryRun = (
  tabId: string,
  sql: string,
  channel: Channel<QueryEvent>,
  limits?: QueryLimits,
) => invoke<void>("query_run", { tabId, sql, channel, limits: limits ?? null });

export const queryCancel = (tabId: string) => invoke<void>("query_cancel", { tabId });

/** Nombre y tipo de cada columna del resultado. */
export interface ColumnType {
  name: string;
  typeName: string;
}

/**
 * Los tipos de las columnas de una sentencia, sin ejecutarla. Cuesta prepararla en el servidor, así
 * que se pide solo cuando el usuario quiere verlos.
 */
export const queryColumnTypes = (tabId: string, sql: string) =>
  invoke<ColumnType[]>("query_column_types", { tabId, sql });

export const queryCommit = (tabId: string) => invoke<TxStatus>("query_commit", { tabId });

export const queryRollback = (tabId: string) => invoke<TxStatus>("query_rollback", { tabId });

/** Encenderlo no confirma la transacción abierta: solo deja de abrir una nueva en cada ejecución. */
export const queryAutocommit = (tabId: string, enabled: boolean) =>
  invoke<TxStatus>("query_autocommit", { tabId, enabled });

export const queryTxStatus = (tabId: string) => invoke<TxStatus>("query_tx_status", { tabId });

export const queryExplain = (tabId: string, sql: string, options?: ExplainOptions) =>
  invoke<Plan>("query_explain", { tabId, sql, options: options ?? null });

export interface RelationColumn {
  name: string;
  typeName: string;
  comment?: string | null;
}

export interface SchemaRelation {
  /** Para el `Ctrl`+clic que revela la tabla en el árbol. */
  oid: number;
  schema: string;
  name: string;
  comment?: string | null;
  columns: RelationColumn[];
}

export interface SchemaSnapshot {
  database: string;
  schemas: string[];
  relations: SchemaRelation[];
}

export const schemaSnapshot = (id: string, database?: string) =>
  invoke<SchemaSnapshot>("schema_snapshot", { id, database: database ?? null });

export interface SqlStatement {
  text: string;
  /** Desplazamiento del primer carácter dentro del script, en caracteres. */
  offset: number;
  line: number;
}

export interface HistoryEntry {
  id: number;
  /** De dónde salió: escrito en el editor, o generado y aplicado por un diálogo. */
  source: "editor" | "dialog";
  profileId: string;
  database: string;
  sql: string;
  /** Segundos desde el epoch. */
  startedAt: number;
  seconds: number;
  rowCount: number | null;
  succeeded: boolean;
  error: string | null;
}

export const historyRecent = (id?: string, limit?: number) =>
  invoke<HistoryEntry[]>("history_recent", { id: id ?? null, limit: limit ?? null });

export const historySearch = (text: string, limit?: number) =>
  invoke<HistoryEntry[]>("history_search", { text, limit: limit ?? null });

export const historyClear = () => invoke<void>("history_clear");

/**
 * Una consulta que el usuario decidió conservar, con nombre. Espejo de `pgforge_core::sql::saved`.
 *
 * Distinta del historial: eso es lo que pasó y se vacía sin que duela; esto es lo que se guarda a
 * propósito. El servidor y la base son de dónde salió, no una atadura: se puede abrir contra otra.
 */
export interface SavedQuery {
  id: number;
  name: string;
  sql: string;
  profileId: string | null;
  database: string | null;
  /** Segundos desde el epoch. */
  createdAt: number;
  updatedAt: number;
}

export const savedList = () => invoke<SavedQuery[]>("saved_list");

/** Con `id` reescribe esa consulta; sin él guarda una nueva. Un nombre repetido devuelve conflicto. */
export const savedSave = (query: {
  id?: number | null;
  name: string;
  sql: string;
  profileId?: string | null;
  database?: string | null;
}) =>
  invoke<SavedQuery>("saved_save", {
    query: {
      id: query.id ?? null,
      name: query.name,
      sql: query.sql,
      profileId: query.profileId ?? null,
      database: query.database ?? null,
    },
  });

export const savedDelete = (savedId: number) => invoke<boolean>("saved_delete", { savedId });

export const statementAtCursor = (sql: string, cursor: number) =>
  invoke<SqlStatement | null>("statement_at_cursor", { sql, cursor });

/**
 * Formatea un script SQL. Nunca falla: `sql::format` es pura y, si no entiende algo, la guarda de
 * equivalencia del núcleo devuelve el texto tal cual llegó en vez de arriesgarse a romperlo.
 */
export const sqlFormat = (sql: string) => invoke<string>("sql_format", { sql });

export const explainWarning = (sql: string, options?: ExplainOptions) =>
  invoke<string | null>("explain_warning", { sql, options: options ?? null });

/** Guarda el texto de una pestaña de consulta como archivo. */
export const sqlWriteFile = (path: string, sql: string) =>
  invoke<void>("sql_write_file", { path, sql });

export const sqlReadFile = (path: string) => invoke<string>("sql_read_file", { path });